File: spike/js/session.js
/*
 * spike/js/session.js - the editor session: selection, collapse, drag, edits.
 *
 * Everything an author's gestures mean, with no DOM anywhere in the file. The
 * shipped editor keeps this state in the one stateful LiveView component
 * (ADR-0005 decision 13) and re-renders after every command; the spike keeps
 * it in a plain value and re-renders after every command, which is the same
 * shape with a different transport. `interact.js` is the translation layer
 * that turns pointer events into calls on this module and the resulting
 * session back into DOM - and it is the ONLY file that knows what an event is.
 *
 * The split is not tidiness. Pre-hover drop validity, the "+" picker's type
 * list, the move-index correction and the undo stacks are all answerable
 * without a browser, so they are all asserted in `dev/selftest.html` against
 * this module rather than probed through a rendered page.
 *
 * ## What this module does NOT own
 *
 *   - The command algebra and its inverses. That is `edit.js`, mirrored from
 *     `StatifierBlocks.Edit`, and every mutation here goes through its
 *     `commit`/`undo`/`redo` funnel. This file builds commands; it never
 *     applies one itself.
 *   - Drop-target enumeration. That is `targets.js`, mirrored from
 *     `Edit.Targets.droppable_slots/3`. A drag session holds the set that
 *     module computed and consults it; it never re-derives a rule.
 *
 * ## The one piece of arithmetic that lives here and nowhere else
 *
 * `moveIndexFor`. ADR-0005 decision 4 fixes the command's meaning - a move's
 * target index is read against the slot with the block ALREADY REMOVED - and
 * `edit.js` implements exactly that. But a gap in the rendered canvas is
 * numbered against the slot as it stands NOW, with the dragged block still in
 * it. Dropping a block on the gap after its two later siblings means index 3
 * to the canvas and index 2 to the command. Converting between the two is a
 * gesture-layer concern, so it is here, where a test can reach it, rather than
 * inside an event handler.
 */

import {
  block,
  err,
  fetchPath,
  findBlock,
  mintBlockId,
  ok,
  subtreeIds,
} from "./document.js";
import {
  canRedo,
  canUndo,
  commit,
  createHistory,
  insert,
  move,
  redo,
  remove,
  target,
  undo,
  updateConfig,
} from "./edit.js";
import { describe, paletteEntryFor } from "./palette.js";
import { droppableSlotSet, slotKey } from "./targets.js";

/* =============================================================== the value */

/**
 * A fresh session over one decoded document.
 *
 *     {
 *       document, registry, history,
 *       selectedId,  the block whose chrome is selected, or null
 *       collapsed,   a Set of block ids folded shut
 *       drag,        the live drag session, or null
 *     }
 *
 * `collapsed` is a Set of IDS rather than of layout nodes on purpose: the
 * canvas is rebuilt from scratch after every command, so anything keyed by a
 * rendered element would be lost on the first edit. Keyed by id, a fold
 * survives an insert, a move, an undo, and a document reload of the same
 * fixture - which is the requirement, stated as a data-structure choice.
 */
export function createSession(document, registry, { history = createHistory() } = {}) {
  return {
    document,
    registry,
    history,
    selectedId: null,
    collapsed: new Set(),
    drag: null,
  };
}

/* ================================================================ selection */

/** Selects `id`, or clears the selection when `id` is `null` or unknown. */
export function select(session, id) {
  if (id !== null && !findBlock(session.document, id)) return deselect(session);
  return session.selectedId === id ? session : { ...session, selectedId: id };
}

export const deselect = (session) => select(session, null);

export const selectedBlock = (session) =>
  session.selectedId === null ? null : findBlock(session.document, session.selectedId);

/**
 * The three fields the inspector's Block section shows. `null` when nothing is
 * selected; the schema-driven config form below it is sb-8cm's subject.
 *
 * The root block reports slot `"root"` rather than an empty string: it is in
 * no slot, and saying so in words beats showing a blank an author has to
 * interpret.
 */
export function selectionSummary(session) {
  const node = selectedBlock(session);
  if (!node) return null;

  const position = positionOf(session.document, node.id);
  const { unresolved } = describe(session.registry, node);

  return {
    id: node.id,
    type: node.type,
    slot: position ? position.slot : "root",
    parentId: position ? position.parentId : null,
    index: position ? position.index : null,
    unresolved,
  };
}

/** Where a block currently sits, or `null` for the root and for absentees. */
export function positionOf(doc, id) {
  const path = fetchPath(doc, id);
  return path === null || path.length === 0 ? null : path[path.length - 1];
}

/* ================================================================= collapse */

/** Folds a block shut, or opens it. Idempotent per call, its own inverse. */
export function toggleCollapsed(session, id) {
  const collapsed = new Set(session.collapsed);

  if (collapsed.has(id)) collapsed.delete(id);
  else collapsed.add(id);

  return { ...session, collapsed };
}

export const isCollapsed = (session, id) => session.collapsed.has(id);

export const expandAll = (session) =>
  session.collapsed.size === 0 ? session : { ...session, collapsed: new Set() };

/* ===================================================================== drag */

/**
 * Starts a drag and computes the valid-target set ONCE, which is the whole
 * point of ADR-0005 decision 5 and the reason every accepting slot can light
 * up before the pointer has been anywhere near it.
 *
 * Two sources, one code path:
 *
 *   `{ kind: "move", id }`      an existing block, relocating
 *   `{ kind: "insert", block }` a block the palette just constructed
 *
 * `targets.js` answers both through `droppableSlotsFor`, so a palette drag and
 * a rearrange drag cannot disagree about what is legal.
 */
export function beginDrag(session, source) {
  const candidate =
    source.kind === "move" ? findBlock(session.document, source.id) : source.block;

  if (!candidate) return session;

  return {
    ...session,
    drag: {
      kind: source.kind,
      id: source.kind === "move" ? source.id : candidate.id,
      label: source.label ?? candidate.type,
      candidate,
      slots: droppableSlotSet(session.document, session.registry, candidate),
    },
  };
}

export const endDrag = (session) => (session.drag ? { ...session, drag: null } : session);

/**
 * What `data-drop` a slot carries right now: `"ok"`, `"no"`, or `null` when no
 * drag is running at all. Three states rather than two, because "no drag" and
 * "this slot refuses the dragged block" must not style the same - a canvas
 * where every slot reads as rejecting at rest is a canvas that looks broken.
 */
export function dropStateFor(session, parentId, slot) {
  if (!session.drag) return null;
  return session.drag.slots.has(slotKey({ parentId, slot })) ? "ok" : "no";
}

export const isDropAllowed = (session, parentId, slot) =>
  dropStateFor(session, parentId, slot) === "ok";

/** How many slots the live drag would accept. The drag banner's counter. */
export const droppableCount = (session) => (session.drag ? session.drag.slots.size : 0);

/* ============================================================ the mutations */

/**
 * Converts a CANVAS gap index into a COMMAND index for a move (see the header
 * comment). Returns `to.index` unchanged for an insert, for a cross-slot move,
 * and for a move to an earlier position in the same slot; only a later
 * position in the same slot shifts, and only by one.
 */
export function moveIndexFor(doc, id, to) {
  const from = positionOf(doc, id);

  const sameSlot = from && from.parentId === to.parentId && from.slot === to.slot;
  return sameSlot && from.index < to.index ? to.index - 1 : to.index;
}

/**
 * Completes the live drag at one gap. Refuses - as a value, never a throw -
 * when there is no drag, or when the gap's slot is not in the set computed at
 * drag start. The second check is not belt-and-braces: a keyboard or a stale
 * pointer can reach a gap the highlight never offered, and decision 5's four
 * rules are the editor's only hard refusal.
 */
export function dropAt(session, to) {
  const drag = session.drag;
  if (!drag) return err({ tag: "no_drag" });

  if (!isDropAllowed(session, to.parentId, to.slot)) {
    return err({ tag: "not_droppable", target: to });
  }

  const command =
    drag.kind === "move"
      ? move(drag.id, target(to.parentId, to.slot, moveIndexFor(session.document, drag.id, to)))
      : insert(target(to.parentId, to.slot, to.index), drag.candidate);

  const applied = runCommand(endDrag(session), command);
  if (!applied.ok) return applied;

  return ok(select(applied.value, drag.id));
}

/**
 * The no-drag insertion path (ADR-0005 decision 8). Same command, same
 * predicate, no drag session involved - which is exactly why the record asks
 * for it: everything a drop can do is reachable from a button.
 */
export function insertTypeAt(session, typeName, to) {
  const node = blockForType(session.registry, typeName);
  if (!node) return err({ tag: "unknown_block_type", type: typeName });

  const staged = beginDrag(session, { kind: "insert", block: node });
  return dropAt(staged, to);
}

/**
 * The types a "+" at this position may offer: every registered type whose
 * freshly-constructed block this slot would accept. It runs decision 5's
 * predicate over each candidate rather than reimplementing the four rules, so
 * the picker and the drag highlight can never disagree.
 */
export function typesForSlot(session, parentId, slot) {
  const offered = [];

  for (const typeName of Object.keys(session.registry.types)) {
    const node = blockForType(session.registry, typeName);
    if (!node) continue;

    const slots = droppableSlotSet(session.document, session.registry, node);
    if (!slots.has(slotKey({ parentId, slot }))) continue;

    const descriptor = session.registry.types[typeName];
    const entry = paletteEntryFor(descriptor);

    offered.push({
      type: typeName,
      label: entry.label,
      description: entry.description,
      group: entry.group,
      icon: entry.icon,
      order: entry.order,
    });
  }

  return offered.sort(
    (a, b) =>
      (a.group < b.group ? -1 : a.group > b.group ? 1 : 0) ||
      a.order - b.order ||
      (a.label < b.label ? -1 : a.label > b.label ? 1 : 0)
  );
}

/**
 * Removes a block and selects what was above it, so an author who deletes four
 * steps in a row is not hunting for the cursor between each one. The root is
 * refused here rather than at `edit.js`'s depth so the refusal can carry the
 * gesture's vocabulary.
 */
export function removeBlock(session, id) {
  if (!id) return err({ tag: "nothing_selected" });
  if (session.document.root.id === id) return err({ tag: "cannot_remove_root", id });

  const position = positionOf(session.document, id);
  const applied = runCommand(endDrag(session), remove(id));
  if (!applied.ok) return applied;

  return ok(select(applied.value, position ? position.parentId : null));
}

/** `updateConfig` through the same funnel, so decision 9's gate always runs. */
export function updateBlockConfig(session, id, config) {
  return runCommand(session, updateConfig(id, config));
}

/* ================================================================ undo/redo */

export const canUndoSession = (session) => canUndo(session.history);
export const canRedoSession = (session) => canRedo(session.history);

export function undoSession(session) {
  return replay(session, undo(session.history, session.registry, session.document));
}

export function redoSession(session) {
  return replay(session, redo(session.history, session.registry, session.document));
}

function replay(session, result) {
  if (!result.ok) return result;

  const next = {
    ...endDrag(session),
    history: result.value.history,
    document: bumpRevision(result.value.document),
  };

  // A block that undo removed cannot stay selected, and a stale id in
  // `selectedId` would put the inspector in a state no gesture could produce.
  return ok(select(next, next.selectedId));
}

/* ================================================================= internals */

/*
 * The one funnel. Every mutation in this file goes through `edit.js`'s
 * `commit`, so the config gate, the inverse capture and the redo-stack clear
 * happen once each rather than once per gesture.
 */
function runCommand(session, command) {
  const result = commit(session.history, session.registry, session.document, command);
  if (!result.ok) return result;

  return ok({
    ...session,
    history: result.value.history,
    document: bumpRevision(result.value.document),
  });
}

/*
 * A revision is a monotonic counter, not a position in the undo stack: undoing
 * an insert takes the document forward to a revision that happens to look like
 * an earlier one. Counting every applied command - forward, undone and redone
 * alike - is what makes "revision 12" mean "twelve edits have been applied",
 * which is the only reading under which two sessions cannot land on the same
 * number holding different bytes.
 */
const bumpRevision = (doc) => ({ ...doc, revision: doc.revision + 1 });

/**
 * A brand-new block of one registered type, its config seeded from the
 * declared defaults in `configSchema/1`.
 *
 * Two passes over the schema, because a schema is a function OF the config -
 * `core.branch` publishes one expression field per arm - so the fields a block
 * declares once it holds its own defaults are not always the fields it
 * declared while empty. Two passes reach the fixed point for every shape the
 * spike's vocabulary has; a type whose schema oscillates would need a real
 * fixed-point loop, and none does.
 */
export function blockForType(registry, typeName) {
  const descriptor = registry.types[typeName];
  if (!descriptor) return null;

  let config = {};

  for (let pass = 0; pass < 2; pass += 1) {
    config = seedConfig(descriptor, config);
  }

  return block({ id: mintBlockId(), type: typeName, typeVersion: descriptor.currentVersion, config });
}

function seedConfig(descriptor, config) {
  const seeded = {};

  for (const field of safeSchema(descriptor, config)) {
    if (field.default === undefined) continue;
    writeValuePath(seeded, field.valuePath ?? [field.key], field.default);
  }

  return seeded;
}

function safeSchema(descriptor, config) {
  try {
    return descriptor.configSchema(config) ?? [];
  } catch {
    return [];
  }
}

/*
 * The write half of `layout.js`'s `readValuePath`: string keys build objects,
 * integer keys build arrays. Written with `defineProperty` at every step for
 * the reason `document.js` gives - a `valuePath` is data, and a `"__proto__"`
 * segment must create an own key rather than move the prototype.
 */
function writeValuePath(root, path, value) {
  let cursor = root;

  for (let index = 0; index < path.length - 1; index += 1) {
    const segment = path[index];
    const existing = Object.hasOwn(cursor, segment) ? cursor[segment] : undefined;
    const child =
      existing !== null && typeof existing === "object"
        ? existing
        : Number.isInteger(path[index + 1])
          ? []
          : {};

    define(cursor, segment, child);
    cursor = child;
  }

  if (path.length > 0) define(cursor, path[path.length - 1], value);

  return root;
}

function define(host, key, value) {
  Object.defineProperty(host, key, {
    value,
    enumerable: true,
    writable: true,
    configurable: true,
  });
}

/**
 * How many blocks a folded subtree is hiding - the collapsed card's count.
 * Built on `subtreeIds`, which is the walk the whole spike counts with, minus
 * the block itself: a folded group reports what is INSIDE it, not what is
 * inside it plus itself.
 */
export function descendantCount(doc, id) {
  const node = findBlock(doc, id);
  return node ? subtreeIds(node).size - 1 : 0;
}

File: lib/statifier_blocks/edit/session.ex
defmodule StatifierBlocks.Edit.Session do
  @moduledoc """
  The commit funnel: one place a gesture reaches the document, over
  `StatifierBlocks.Edit.History.commit/4`.

  A session is the editing state a surface holds between gestures - the
  palette it resolves types through, the document it is editing, the undo
  history behind it, the per-block config drafts the document has refused,
  and the last refusal that had nowhere else to go. Every function here is
  a pure function of that value: it takes a session and returns a session,
  with no socket, no assigns and no process state in it.

  ## Why the funnel is a module and not four copies

  The package's own editor and every host embedding it face the same four
  lines of `Edit.History.commit/4` handling, and the decisions inside them
  are the package's, not the surface's:

    * a refused `{:update_config, id, config}` is held as a **draft** of
      what the author typed, keyed by block id, rather than being blanked
      back to the document's value (ADR-0005 decision 9). A host that
      re-implements the funnel re-implements that decision, and gets it
      wrong by keeping the keystrokes it should drop or dropping the ones
      it should keep;
    * a refusal that is not about a config lands in `last_error` and moves
      nothing else;
    * an undo or a redo drops **every** draft: a draft is by definition
      config the document never accepted, and carrying one across a
      history move shows the author a value belonging to a document state
      they just stepped out of;
    * a list gesture reads and writes through the field's own `value_path`
      (ADR-0002 decision 7, amended 2026-08-27) rather than through its
      key, and descends into nested members the same way the control that
      draws them does.

  ## The `{:ok, session}` / `{:error, session}` result

  Every funnel function answers with a tagged session, and the tag says
  exactly one thing: **did the document move**. `{:ok, session}` means it
  did, and is the caller's cue to run whatever it does on a change - the
  package editor notifies its `on_change` host callback, a host store
  persists. `{:error, session}` means it did not, and the session still
  comes back because a refusal is a state change too: it is where the
  draft was recorded, or where `last_error` was set. There is nothing for
  a caller to do with it but render it.

  ## Where this module is, and is not

  It lives beside `StatifierBlocks.Edit.History` and
  `StatifierBlocks.Edit.Targets` because it operates on exactly what they
  operate on - a document, a palette and a history - and because, like
  them, it compiles and runs with no LiveView in the tree. The alternative
  home, under `StatifierBlocks.Editor`, is the LiveView component's own
  namespace: ADR-0005 decision 1 compiles it out entirely when LiveView is
  absent, and it is where the socket lives. The socket stays there; the
  document, the history and the drafts come here.

  ## Building one

  A session is an ordinary struct with no constructor to learn:

      %StatifierBlocks.Edit.Session{
        palette: palette,
        document: document,
        history: StatifierBlocks.Edit.History.new()
      }

  `drafts` defaults to `%{}` and `last_error` to `nil`.
  """

  alias StatifierBlocks.{Block, BlockType, Document, Edit, Palette, ViewModel}
  alias StatifierBlocks.Edit.History

  @typedoc "Per-block config the document refused, keyed by block id."
  @type drafts :: %{optional(Block.id()) => Block.config()}

  @type t :: %__MODULE__{
          palette: Palette.t(),
          document: Document.t(),
          history: History.t(),
          drafts: drafts(),
          last_error: term() | nil
        }

  @enforce_keys [:palette, :document, :history]
  defstruct [:palette, :document, :history, drafts: %{}, last_error: nil]

  @typedoc "What a list control asks of one member list: append a blank, or drop the member at an index."
  @type gesture :: :add | {:remove, integer()}

  @typedoc """
  A gesture, optionally addressed at a nested member list.

  The bare form is the field's own list. A `{path, gesture}` pair steps
  into that list's member at each index in turn - the same descent the
  control makes when it nests - and applies the gesture there.
  """
  @type list_gesture :: gesture() | {[non_neg_integer()], gesture()}

  @doc """
  Applies `command` to the session's document through the gated history.

  The one place a command reaches the document, so the gate, the undo
  stack and the caller's change notification each have one implementation
  rather than one per gesture.

  On `{:ok, session}` the document and the history have both moved and
  `last_error` is cleared; drafts are untouched, because a command that is
  not a config change says nothing about a config the document refused. On
  `{:error, session}` the refusal is in `last_error` and nothing else
  moved.

  ## Examples

      iex> alias StatifierBlocks.{Block, Document, Palette}
      iex> alias StatifierBlocks.Edit.{History, Session}
      iex> document = Document.new(Block.new("core.sequence", id: "blk_ROOT", slots: %{"body" => []}))
      iex> session = %Session{palette: Palette.core(), document: document, history: History.new()}
      iex> {:ok, block} = Palette.new_block(session.palette, "core.placeholder")
      iex> {:ok, moved} = Session.commit(session, {:insert, {"blk_ROOT", "body", 0}, block})
      iex> length(Document.blocks(moved.document))
      2
      iex> {:error, refused} = Session.commit(session, {:remove, "blk_NOPE"})
      iex> refused.last_error
      {:no_such_block, "blk_NOPE"}
  """
  @spec commit(t(), Edit.t()) :: {:ok, t()} | {:error, t()}
  def commit(%__MODULE__{} = session, command) do
    case History.commit(session.history, session.palette, session.document, command) do
      {:ok, history, document} ->
        {:ok, %{session | history: history, document: document, last_error: nil}}

      {:error, reason} ->
        {:error, %{session | last_error: reason}}
    end
  end

  @doc """
  Writes `config` onto the block `id`, holding a refused config as a draft.

  This differs from `commit/2` in one way and only one: an
  `{:invalid_config, id, findings}` refusal is recorded in `drafts` under
  `id` instead of in `last_error`. That is ADR-0005 decision 9's draft
  treatment - a value the document refuses is still the value the author
  is holding, and blanking it back to the document's would delete their
  keystrokes to punish a typo. Every other refusal lands in `last_error`
  exactly as `commit/2`'s does.

  A config the document accepts clears that block's draft: the author has
  resolved the refusal that produced it.

  """
  @spec change_config(t(), Block.id(), Block.config()) :: {:ok, t()} | {:error, t()}
  def change_config(%__MODULE__{} = session, id, config) do
    case History.commit(
           session.history,
           session.palette,
           session.document,
           {:update_config, id, config}
         ) do
      {:ok, history, document} ->
        {:ok,
         %{
           session
           | history: history,
             document: document,
             drafts: Map.delete(session.drafts, id),
             last_error: nil
         }}

      {:error, {:invalid_config, ^id, _findings}} ->
        {:error, %{session | drafts: Map.put(session.drafts, id, config)}}

      {:error, reason} ->
        {:error, %{session | last_error: reason}}
    end
  end

  @doc """
  Steps the history one move: `:undo` or `:redo`.

  Both directions arrive at the same reconciliation, because `History` has
  already done the part that differs by the time this runs. Drafts are
  dropped wholesale on a successful step - see the moduledoc for why - and
  an empty stack answers `{:error, session}` with `:nothing_to_undo` or
  `:nothing_to_redo` in `last_error`.

  ## Examples

      iex> alias StatifierBlocks.{Block, Document, Palette}
      iex> alias StatifierBlocks.Edit.{History, Session}
      iex> document = Document.new(Block.new("core.sequence", id: "blk_ROOT", slots: %{"body" => []}))
      iex> session = %Session{palette: Palette.core(), document: document, history: History.new()}
      iex> {:error, nothing} = Session.step(session, :undo)
      iex> nothing.last_error
      :nothing_to_undo
      iex> {:ok, block} = Palette.new_block(session.palette, "core.placeholder")
      iex> {:ok, moved} = Session.commit(session, {:insert, {"blk_ROOT", "body", 0}, block})
      iex> {:ok, back} = Session.step(moved, :undo)
      iex> length(Document.blocks(back.document))
      1
  """
  @spec step(t(), :undo | :redo) :: {:ok, t()} | {:error, t()}
  def step(%__MODULE__{} = session, direction) when direction in [:undo, :redo] do
    stepped =
      case direction do
        :undo -> History.undo(session.history, session.palette, session.document)
        :redo -> History.redo(session.history, session.palette, session.document)
      end

    case stepped do
      {:ok, history, document} ->
        {:ok, %{session | history: history, document: document, drafts: %{}, last_error: nil}}

      {:error, reason} ->
        {:error, %{session | last_error: reason}}
    end
  end

  @doc """
  Applies a list gesture to `field` on the block `id`, and commits it.

  The rows are read from the block's **effective** config - the draft if
  one is held, the document's otherwise - and written back through the
  field's own `value_path/1`, which is where the form's other writes go.
  The commit is `change_config/3`'s, so a list edit that leaves the config
  invalid is held as a draft like any other refused config rather than
  being lost.

  `gesture` is a `t:list_gesture/0`: `:add`, `{:remove, index}`, or either
  of those addressed at a nested member list by a `{path, gesture}` pair.

  """
  @spec update_list(t(), Block.id(), ViewModel.Field.t(), list_gesture()) ::
          {:ok, t()} | {:error, t()}
  def update_list(%__MODULE__{} = session, id, %ViewModel.Field{} = field, gesture) do
    config = Document.effective_config(session.document, id, session.drafts)
    path = ViewModel.Field.value_path(field)

    held =
      case BlockType.fetch_value(config, path) do
        {:ok, value} -> value
        :error -> nil
      end

    rows = apply_gesture(%{field | value: held}, gesture)

    change_config(session, id, BlockType.put_value(config, path, rows))
  end

  @doc """
  The rows `field` would hold after `gesture`, as a value.

  The document is not consulted and nothing is committed: this is the list
  half of `update_list/4` on its own, for a surface that needs the rows
  before it has somewhere to put them. The field's `type` says what a
  blank member looks like and how a stored value reads as rows, and its
  `value` is what the gesture is applied to.

  A `{:list, t}` wraps whatever it finds, which is what a stored scalar
  has always meant there. A `{:type_expr, opts}` does not: its inline arm
  is a list and its name arm is a string, and wrapping the string would
  turn a type name into a nameless member.

  ## Examples

      iex> alias StatifierBlocks.Edit.Session
      iex> alias StatifierBlocks.ViewModel.Field
      iex> field = %Field{key: "tags", type: {:list, :string}, label: "Tags", required?: false, default: [], value: ["a", "b"]}
      iex> Session.apply_gesture(field, :add)
      ["a", "b", ""]
      iex> Session.apply_gesture(field, {:remove, 0})
      ["b"]
  """
  @spec apply_gesture(ViewModel.Field.t(), list_gesture()) :: [term()]
  def apply_gesture(%ViewModel.Field{type: type, value: value}, gesture) do
    {path, bare} = addressed(gesture)

    type
    |> rows_of(value)
    |> descend(path, bare, blank_row(type))
  end

  @spec addressed(list_gesture()) :: {[non_neg_integer()], gesture()}
  defp addressed({path, gesture}) when is_list(path), do: {path, gesture}
  defp addressed(gesture), do: {[], gesture}

  @spec rows_of(BlockType.field_type(), term()) :: [term()]
  defp rows_of({:type_expr, _opts}, value) when is_list(value), do: value
  defp rows_of({:type_expr, _opts}, _not_a_member_list), do: []
  defp rows_of(_type, value), do: List.wrap(value)

  @spec blank_row(BlockType.field_type()) :: term()
  defp blank_row({:type_expr, _opts}), do: %{"name" => "", "type" => "", "required?" => false}
  defp blank_row(_type), do: ""

  @spec descend([term()], [non_neg_integer()], gesture(), term()) :: [term()]
  defp descend(rows, [], :add, blank), do: rows ++ [blank]
  defp descend(rows, [], {:remove, index}, _blank), do: List.delete_at(rows, index)

  defp descend(rows, [index | rest], gesture, blank) do
    List.update_at(rows, index, fn row ->
      inner =
        case row do
          %{"type" => members} when is_list(members) -> members
          _no_nested_members -> []
        end

      row
      |> member_row()
      |> Map.put("type", descend(inner, rest, gesture, blank))
    end)
  end

  @spec member_row(term()) :: map()
  defp member_row(row) when is_map(row), do: row
  defp member_row(_row), do: %{"name" => "", "required?" => false}
end
